src/Features/Core/Portable/ExtractInterface/AbstractExtractInterfaceService.cs

```diff
@@ -177,20 +214,65 @@ public ExtractInterfaceResult ExtractInterfaceFromAnalyzedType(ExtractInterfaceT
+            var typeDeclaration = originalRoot.GetAnnotatedNodes(typeNodeSyntaxAnnotation).Single();
+
+            var trackedDocument = document.WithSyntaxRoot(originalRoot.TrackNodes(typeDeclaration));
+
+            var currentRoot = trackedDocument.GetSyntaxRootSynchronously(cancellationToken);
+            var editor = new SyntaxEditor(currentRoot, annotatedSolution.Workspace);
+
+            // Generate the interface syntax node, which will be inserted above the type it's extracted from
+            var codeGenService = trackedDocument.GetLanguageService<ICodeGenerationService>();
+            var interfaceNode = codeGenService.CreateNamedTypeDeclaration(extractedInterfaceSymbol).WithAdditionalAnnotations(SimplificationHelpers.SimplifyModuleNameAnnotation);
+
+            typeDeclaration = currentRoot.GetCurrentNode(typeDeclaration);
+            editor.InsertBefore(typeDeclaration, interfaceNode);
+
+            var unformattedSolution = document.WithSyntaxRoot(editor.GetChangedRoot()).Project.Solution;
+
+            // After the interface is inserted, update the original type to show it implements the new interface
+            var completedSolution = GetSolutionWithOriginalTypeUpdated(
+                unformattedSolution, documentIds,
+                refactoringResult.DocumentToExtractFrom.Id, typeNodeSyntaxAnnotation,
+                refactoringResult.TypeToExtractFrom, extractedInterfaceSymbol,
+                extractInterfaceOptions.IncludedMembers, symbolToDeclarationAnnotationMap, cancellationToken);
+
+            return new ExtractInterfaceResult(
+                succeeded: true,
+                updatedSolution: completedSolution,
+                navigationDocumentId: refactoringResult.DocumentToExtractFrom.Id);
+        }
+
+        private 
+            (Dictionary<ISymbol, SyntaxAnnotation> symbolToDeclarationAnnotationMap, 
+            Solution annotatedSolution, 
+            List<DocumentId> documentIds, 
+            SyntaxAnnotation typeNodeAnnotation) 
+            CreateSymbolToDeclarationAnnotationMap(
+                IEnumerable<ISymbol> includedMembers,
+                Solution solution,
+                SyntaxNode typeNode,
+                CancellationToken cancellationToken)
         {
             var symbolToDeclarationAnnotationMap = new Dictionary<ISymbol, SyntaxAnnotation>();
             var currentRoots = new Dictionary<SyntaxTree, SyntaxNode>();
-            documentIds = new List<DocumentId>();
+            var documentIds = new List<DocumentId>();
 
             var typeNodeRoot = typeNode.SyntaxTree.GetRoot(CancellationToken.None);
-            typeNodeAnnotation = new SyntaxAnnotation();
+            var typeNodeAnnotation = new SyntaxAnnotation();
             currentRoots[typeNode.SyntaxTree] = typeNodeRoot.ReplaceNode(typeNode, typeNode.WithAdditionalAnnotations(typeNodeAnnotation));
             documentIds.Add(solution.GetDocument(typeNode.SyntaxTree).Id);
 
```

Comment:
I am not sure why code below that gets tree from location use GetRoot to get node. it should use GetRootAsync

Reply:
okay. so what location pointing to in this case? name node or decl node? is broken case like empty name case are all filtered out already I guess?